### PR TITLE
[WIP] Better doxygen for PROPERTY's

### PR DESCRIPTION
**Pull request by chrisdembia**
The way we document properties is by showing the macro invocating in doxygen. Ideally we would show in doxygen the methods that the macros define. Here's an attempt at doing so!

image

View the complete doxygen result here:
http://stanford.edu/~dembia/opensim-better-property-docs/

This work is not done yet, but I would like feedback:

1. Should we show ALL the methods that are created, for completeness? I feared that would be too cluttered.
2. Should we show the comment for ALL methods for each property, or just the getter, as shown?
3. Should we get rid of the "Property declarations" group, and create a separate group for each individual property, and have the comment for the property go into the description of that group?
4. Ideally we would get rid of the double quotes in the comments, but IDK how to do that. Any ideas? We could remove the quotes from the input to the macro, and use `#comment` inside the macro, or something like that? This is undesirable though.

@aseth1 mentioned trying to use in-class member initializers for properties in the future, which would make this PR unnecessary.

TODO:

- [ ] Replace instances of "Property declarations" with "Properties"
- [x] Add extra description within the macros describing where to find more information.
- [ ] Check each type of PROPERTY macro; ensure they all appear nicely now.

### CODE REVIEW

**Comment by chrisdembia** [comment]
Add description to the revealed "getters" saying that "these are autogenerated", and list the other methods that are created.

**Comment by chrisdembia** [comment]
Based on feedback from a previous opensim meeting, I've updated this doxygen output. The new doxygen is here: http://stanford.edu/~dembia/opensim-better-property-docs-20150227/

I now only show ONE method for each property; its getter. In the description for the getter, I include more extensive documentation (reference to other methods, to Property, and to the macro used to generate the property). Here are a few screenshots:

### Property declarations for Model
image

### Property declarations for Coordinate
image

### OpenSim_DECLARE_PROPERTY
image

### OpenSim_DECLARE_UNNAMED_PROPERTY
image

### OpenSim_DECLARE_OPTIONAL_PROPERTY
image

### OpenSim_DECLARE_LIST_PROPERTY
image

### OpenSim_DECLARE_LIST_PROPERTY_SIZE
image

### Discussion
One downside is that the doxygen for the macros themselves now have lots of bad-looking slashes in them, due to the doxygen comments within the macros. I can fix this by putting the comments in their own macros.
image

Ideally, the setter, getter, etc. would not appear at the top under the method declarations, but would still appear at the bottom of each doxygen page. That way the autogenerated methods would still have a link and appear to actually exist. However, I don't know how to do this; I created a [stackoverflow post](http://stackoverflow.com/questions/28754115/in-doxygen-how-do-i-hide-a-member-function-from-public-member-functions-but#comment45795497_28754115) about this.

Also, I'm not necessarily convinced that this approach is the way to go. Perhaps we should just go back to the original macros, but add more description within the macro's doxygen? Please let me know! I definitely think it is important to improve the visibility of these auto-generated methods. I know at least personally I didn't want to read the thorough description at Property.h, even though it's very well written and clear, when just looking for the methods available in one class.

/cc @sherm1 @jimmyDunne @tkuchida @aymanhab @aseth1 @carmichaelong

**Comment by chrisdembia** [comment]
Feedback from dev meeting: try creating a new separate group for the upd, set, etc methods; remove the type from the brief description.

**Comment by chrisdembia** [comment]
Look at this!

doxygen

Isn't that snazzy? Doxygen supports Qt properties, and I made Doxygen think that our properties are Qt properties.

**Comment by jimmyDunne** [comment]
wooooo, looks great!

**Comment by aseth1** [comment]
Beautiful!
The "required" and "unnamed" distinctions somewhat confusing. All the required ones have default values, so if you don't set them you are OK. Being named or unnamed is likely not to be helpful to users. I would be happy if everything before the ':' disappeared in the description.

Also, is it straightforward to add a generic block that lists the methods associated with each property (e.g. get_<p>(),  set_<p>(), ....)? Or should we just link to the description of the Property class for those?

**Comment by chrisdembia** [comment]
> The "required" and "unnamed" distinctions somewhat confusing. All the required ones have default values, so if you don't set them you are OK. Being named or unnamed is likely not to be helpful to users. I would be happy if everything before the ':' disappeared in the description.

Okay good point. I'll still leave the `See also` to the macro that was used to generate the property.



> Also, is it straightforward to add a generic block that lists the methods associated with each property (e.g. get_<p>(), set_<p>(), ....)?

Yes I'll do this.

**Comment by sherm1** [comment]
Wow! Great idea, Chris!

**Comment by aseth1** [comment]
@opensim-org has everyone see this?

**Comment by aseth1** [comment]
@opensim-org has everyone see this?

**Comment by chrisdembia** [comment]
Hey, I have an exciting update for this PR and I'm interested in getting feedback and getting this merged soon.

[**Check out the new doxygen proposal here**](http://stanford.edu/~dembia/opensim-better-property-docs-20150522/classOpenSim_1_1Model.html)

1. Properties now look like this:
image
2. For each property, there is a detailed description at the bottom of the class page:
image
3. Here's a list property's detailed description. It includes the range of possible lengths for the list
image
4. I've also exposed a subset of the macro-generated methods. They appear underneath the Public Member Functions, and they also have detailed descriptions
image
image
4. The organization of the class page: Properties are near top of the page (below inheritance diagram but above public member functions). Then the list of property-related methods that are generated by the property macros are the *first* named group within Public Member Functions:
image
5. Protected or private properties. The property itself is shown even for protected properties. I think this is good, since they are also exposed in XML. If you see a tag in XML files you should be able to see it in doxygen. **However**, the methods generated by the protected property end up in the protected section. In Component, the `_components` property is protected:
image
image
6. A **really**  neat side effect is now there is a dropdown list of **inherited** properties from your superclasses!! Here are clips from the Millard muscle:
image
image

6. To get the organization of the page to look right, we need to remove the `/** @name Property delcarations` block from all header files (which is a good thing). Right now, you should focus on the Model and Muscle classes, because for these classes I already removed these blocks. For all other classes, the organization of the class page is messed up.

As a sidenote, I am exploiting Doxygen's ability to document Qt `Q_PROPERTY` macros. To allow our users to still use Qt with OpenSim, I don't *actually* use the `Q_PROPERTY` macro; I use an `OpenSim_FAKE_Q_PROPERTY` that only doxygen defines as `Q_PROPERTY` in its preprocessor.

If you are satisfied with the result, then I'll go through the code base and remove the `/** @name Property declarations` lines that we have in all our header files. Then I'll open a separate and clean PR with the complete changes.


/cc @aseth1 @sherm1 @carmichaelong @jimmyDunne @jenhicks @klshrinidhi @aymanhab

File: OpenSim/Common/Property.h [hunk 1]
```diff
@@ -1121,16 +1128,30 @@ SimTK_DEFINE_UNIQUE_INDEX_TYPE(PropertyIndex);
...
+// to nicely document OpenSim properties. See, Doxygen has support for Qt
+// "properties", which are slightly similar to our properties. Qt "properties"
+// are defined with the Q_PROPERTY() macro. In the code, this macro is empty
+// (as you can tell by...just looking at the macro).  But in the Doxyfile, we
+// tell Doxygen that this macro is defined as Q_PROPERTY(). This tricks Doxygen
+// into documenting *our* properties as if they were Qt properties.
+// The reason that we don't use Q_PROPERTY() directly is that we don't want to
+// prevent our users from potentially using OpenSim in conjunction with Qt.
+// You can see how this empty macro is used in some of the macros below.
+#define OpenSim_FAKE_Q_PROPERTY(T, name)
```
Comment (tkuchida) [inline OpenSim/Common/Property.h]: Change `FAKE` to `DOXYGEN`?

**Comment by tkuchida** [comment]
Looks great! I *really* like the lists of inherited properties.

File: OpenSim/Common/Property.h (hunk 1 above)
Comment (chrisdembia) [inline OpenSim/Common/Property.h]: +1

**Comment by chrisdembia** [comment]
> Looks great! I really like the lists of inherited properties.

You mentioned earlier that there are just way too many methods to look at as a result. Do you still think so? If so, do you think we should hide all the generated methods? We could leave the Properties section and the detailed description of each property could just list the generated methods.

**Comment by tkuchida** [comment]
>do you think we should hide all the generated methods?

The "Property-related methods" section is verbose, but it's easy to skip over once you get the hang of how the property system works. I would leave it the way it is.

**Comment by sherm1** [comment]
> Looks great! I really like the lists of inherited properties.

Me too! :+1:

**Comment by jenhicks** [comment]
Thanks Chris! This looks great to me. My only question is how the subset you mention in (4) is determined. 

I think this is good to merge. It is a big improvement over the current doxygen and nothing will keep us from further iterations to improve.

**Comment by chrisdembia** [comment]
> My only question is how the subset you mention in (4) is determined.

I show only (and all) the methods that scripting users should use. There are additional methods that are intended for developers (`constructProperty_propname()`); in the future we can potentially reveal all auto-generated methods in the developer doxygen.

Thanks for looking at this.

**Comment by jenhicks** [comment]
Sounds good -- thanks for clarifying

**Comment by chrisdembia** [comment]
I'll close this PR and open a new one in which I remove the `@name Property declarations` lines.

**Pull request closed without merging**